File: service/appstore.go
/*
 * [INPUT]: 依赖内嵌的默认 App Store 清单与用户可覆盖的 <data-dir>/appstore.json
 * [OUTPUT]: 对外提供 StoreApp 目录（名称、类型、GitHub repository），供 Agent 经 recut.apps.store 发现可安装 App
 * [POS]: service 的应用分发发现边界；商店清单可被本地文件覆盖，使社区/用户能扩展可安装目录而不改代码
 * [PROTOCOL]: 变更时更新此头部，然后检查 README.md
 */
package main

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

//go:embed appstore/apps.json
var embeddedAppStore []byte

// StoreApp is one installable App advertised by the store.
type StoreApp struct {
	AppID       string                       `json:"appId"`
	Name        string                       `json:"name"`
	Description string                       `json:"description"`
	Kind        string                       `json:"kind"`
	Repository  string                       `json:"repository"`
	Localized   map[string]StoreAppLocalized `json:"localized,omitempty"`
}

// StoreAppLocalized is the per-locale override for a store entry's name and
// description; empty values keep the top-level (author) language.
type StoreAppLocalized struct {
	Name        string `json:"name,omitempty"`
	Description string `json:"description,omitempty"`
}

// AppStore returns the store catalog in the default (zh) locale.
func (s *Store) AppStore() ([]StoreApp, error) {
	return s.AppStoreFor(DefaultLocale)
}

// AppStoreFor returns the store catalog with each entry's name and description
// merged against the requested locale's localized block. The user-provided
// <data-dir>/appstore.json wins over the embedded default, exactly like
// AppStore; the merge is applied after decoding.
func (s *Store) AppStoreFor(locale Locale) ([]StoreApp, error) {
	data := embeddedAppStore
	if override, err := os.ReadFile(filepath.Join(s.root, "appstore.json")); err == nil {
		data = override
	} else if !os.IsNotExist(err) {
		return nil, err
	}
	apps := []StoreApp{}
	if err := json.Unmarshal(data, &apps); err != nil {
		return nil, fmt.Errorf("parse app store catalog: %w", err)
	}
	for index := range apps {
		if apps[index].AppID == "" || apps[index].Name == "" || apps[index].Repository == "" {
			return nil, fmt.Errorf("app store entry %d is missing appId, name, or repository", index)
		}
		if localized, ok := apps[index].Localized[string(locale)]; ok {
			if localized.Name != "" {
				apps[index].Name = localized.Name
			}
			if localized.Description != "" {
				apps[index].Description = localized.Description
			}
		}
	}
	return apps, nil
}

File: web/lib/appstore.ts
/*
 * [INPUT]: 依赖构建时的云端市场基址（NEXT_PUBLIC_RECUT_SITE_URL，默认 https://recut.video）与 locales.ts 的 Locale
 * [OUTPUT]: 应用市场（可添加 App）的唯一云端数据源：从 <site>/api/appstore.json 拉取双语言目录，按当前语言取 name/description/requirements；离线或拉取失败返回空目录
 * [POS]: web/lib 的应用市场云端边界；工作台 Apps 市场与 App 详情页消费，取代静态 app-catalog；service 的嵌入式 appstore 仅服务 MCP recut.apps.store 发现
 * [PROTOCOL]: 变更时更新此头部，然后检查 README.md
 */
import { type Locale, defaultLocale } from "@/lib/i18n/locales";

export const marketplaceBase = process.env.NEXT_PUBLIC_RECUT_SITE_URL ?? "https://recut.video";

export type MarketplaceRequirements = { title: string; items: string[]; note?: string };

export type MarketplaceApp = {
  appId: string;
  author: string;
  kind: "project" | "standalone";
  repository: string;
  name: Record<Locale, string>;
  description: Record<Locale, string>;
  requirements?: Record<Locale, MarketplaceRequirements>;
};

export function marketplaceName(app: MarketplaceApp, locale: Locale): string {
  return app.name[locale] ?? app.name[defaultLocale] ?? app.name.zh;
}

export function marketplaceDescription(app: MarketplaceApp, locale: Locale): string {
  return app.description[locale] ?? app.description[defaultLocale] ?? app.description.zh;
}

export function marketplaceRequirements(app: MarketplaceApp, locale: Locale): MarketplaceRequirements | undefined {
  const requirements = app.requirements?.[locale] ?? app.requirements?.zh;
  return requirements && (requirements.title !== "" || requirements.items.length > 0) ? requirements : undefined;
}

export async function loadMarketplace(): Promise<MarketplaceApp[]> {
  try {
    const response = await fetch(`${marketplaceBase}/api/appstore.json`, { cache: "no-store" });
    if (!response.ok) return [];
    const data = (await response.json()) as { apps?: MarketplaceApp[] };
    return data.apps ?? [];
  } catch {
    return [];
  }
}
